[HOTFIX #11072] Introduce the cfb_shell sample app

diegosueiro opened the pull request:
This PR is to fix the compile errors introduced by #11072.

It undoes the reverts 64aecd0, f8625ce, 4e2c9d1, 40eb3a5 and fixes the commit 2e27113

diegosueiro committed:
subsys/fb/cfb: Introduce cfb_framebuffer_invert function

Adds the cfb_framebuffer_invert function to invert Pixels in the
Character Framebuffer

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
subsys/fb/cfb: Introduce cfb_get_numof_fonts function

Adds the cfb_get_numof_fonts function to get the number of
available fonts

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
samples/display: Introduce the cfb_shell sample app

This adds the shell sample app for the Character Framebuffer testing.

cfb - Character Framebuffer shell commands
Options:
  -h, --help  :Show command help.
Subcommands:
  init        :[none]
  get_device  :[none]
  get_param   :<all, height, width, ppt, rows, cols>
  get_fonts   :[none]
  set_font    :<idx>
  invert      :[none]
  print       :<col: pos> <row: pos> <text>
  scroll      :<dir: (vertical|horizontal)> <col: pos> <row: pos> <text>
  clear       :[none]

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@jarz-nordic,

Can you please review this PR as well?

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+		return -ENODEV;
+	}
+
+	if (!strcmp(argv[1], "vertical")) {
```
It is not bad implementation and you can keep it as it is.

From my experience I would use a different convention. 
I would create a subcommands `vertical` and `horizontal` and I would implement handlers there. It will simplify usage from end user perspective. User will be able to prompt / autocomplete them. In this scenario handler `cmd_scroll` would be not needed at all.

Another solution / compromise would be adding subcommands `vertical` and `horizontal` without a handler. User will be able to autocomplete them with the `tab` key but handler `cmd_scroll` could stay as it is.

diegosueiro replied:
Can you please point me some code examples using the subcommands?

jarz-nordic replied:
Well, you are using subcommands in this example, you create them with macro:  `SHELL_CREATE_STATIC_SUBCMD_SET` :)

You can change you have defined:

`SHELL_CMD_ARG(scroll, NULL, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

you can change to (I've added **&sub_scroll**):

`SHELL_CMD_ARG(scroll, &sub_scroll, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

Next you need to create  **&sub_scroll** like that:

```
SHELL_CREATE_STATIC_SUBCMD_SET(sub_scroll) {
    SHELL_CMD_ARG(horizontal, NULL, HELP_NONE, NULL),
    SHELL_CMD_ARG(vertical, NULL, HELP_NONE, NULL),
    SHELL_SUBCMD_SET_END
};
```

It is important that `SHELL_CREATE_STATIC_SUBCMD_SET(sub_scroll)` is defined **above** : `SHELL_CMD_ARG(scroll, &sub_scroll, "<dir: (vertical|horizontal)> " "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),`

**EXAMPLE:**
You can find in file `subsys/shell/shell_cmds.c`
Take a look at command `shell`. You can call it `shell colors off`:

- shell - root command
- colors - subcommand of shell command
- off  - subcommand of colors command

diegosueiro replied:
Understood.
But when calling the `cmd_scroll` I still need to know if it is `vertical` or `horizontal` for example.
If I don't have a handler for each subcommand and use the `cmd_scroll` I still need to use the `strcmp`. Right?

Maybe instead of having a  `cmd_scroll` I add a  `sub_cmd_scroll_vert` and  `sub_cmd_scroll_horz`.

What do you say?

jarz-nordic replied:
Yes, if you wont implement handlers for `vertical` and `horizontal` you will still need to use `strcmp`. However you will be able to complete / prompt them.

I am usually implementing handlers (like for colors subcommands: `on`, and `off')

jarz-nordic replied:
> Maybe instead of having a cmd_scroll I add a sub_cmd_scroll_vert and sub_cmd_scroll_horz.

This is great idea that I support ;)

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+		return -ENODEV;
+	}
+
+	if (!strcmp(param_filter, "all")) {
```
Similar situation. These all can be subcommands with a handler - you won't need `strcmp` and `cmd_get_param` or
these can be subcommands without a handler what will simplify shell usage.

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+}
+
+SHELL_CREATE_STATIC_SUBCMD_SET(cfb_cmds) {
+	SHELL_CMD(init, NULL, HELP_NONE, cmd_init),
```
This could be used with `_ARG`, shell will check arguments count.

`SHELL_CMD_ARG(init, NULL, HELP_NONE, cmd_init, 1, 0),`

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+
+SHELL_CREATE_STATIC_SUBCMD_SET(cfb_cmds) {
+	SHELL_CMD(init, NULL, HELP_NONE, cmd_init),
+	SHELL_CMD(get_device, NULL, HELP_NONE, cmd_get_device),
```
`SHELL_CMD_ARG(get_device, NULL, HELP_NONE, cmd_get_device, 1, 0),`

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+	SHELL_CMD(get_device, NULL, HELP_NONE, cmd_get_device),
+	SHELL_CMD_ARG(get_param, NULL, "<all, height, width, ppt, rows, cols>",
+		      cmd_get_param, 2, 0),
+	SHELL_CMD(get_fonts, NULL, HELP_NONE, cmd_get_fonts),
```
`SHELL_CMD_ARG(get_fonts, NULL, HELP_NONE, cmd_get_fonts, 1, 0),`

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+		      cmd_get_param, 2, 0),
+	SHELL_CMD(get_fonts, NULL, HELP_NONE, cmd_get_fonts),
+	SHELL_CMD_ARG(set_font, NULL, "<idx>", cmd_set_font, 2, 0),
+	SHELL_CMD(invert, NULL, HELP_NONE, cmd_invert),
```
`SHELL_CMD_ARG(invert, NULL, HELP_NONE, cmd_invert, 1, 0),`

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,355 @@
+		      0),
+	SHELL_CMD_ARG(scroll, NULL, "<dir: (vertical|horizontal)> "
+		      "<col: pos> <row: pos> <text>", cmd_scroll, 5, 0),
+	SHELL_CMD(clear, NULL, HELP_NONE, cmd_clear),
```
`SHELL_CMD_ARG(clear, NULL, HELP_NONE, cmd_clear, 1, 0),`

jakub-uC reviewed (commented):
Looks good, I have few proposals which will simplify shell usage.

jarz-nordic commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,470 @@
+	SHELL_CMD_ARG(set_font, NULL, "<idx>", cmd_set_font, 2, 0),
+	SHELL_CMD_ARG(invert, NULL, HELP_NONE, cmd_invert, 1, 0),
+	SHELL_CMD_ARG(print, NULL, HELP_PRINT, cmd_print, 4, 0),
+	SHELL_CMD(scroll, &sub_cmd_scroll, "<dir: (vertical|horizontal)> "
```
I think that you might consider changing help message a bit, it seems that it is duplicated.
Please call `scroll -h` to see what I mean. You can put here a simple description what `scroll` is resposible for, like scrolling what :).

diegosueiro replied:
Addressed.

diegosueiro committed:
shell: Add initial code for Character Framebuffer shell module

This adds the config option and files of the Character Framebuffer
shell module

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

jakub-uC reviewed (approved):
From shell perspective looks good.
